lib/run_loop/physical_device/life_cycle.rb

```diff
@@ -0,0 +1,260 @@
+        raise RunLoop::Abstract::AbstractMethodError,
+          "Subclass must implement '.tool_is_installed?'"
+      end
+
+      # Path to tool.
+      def self.executable_path
+        raise RunLoop::Abstract::AbstractMethodError,
+          "Subclass must implement '.executable_path'"
+      end
+
+      # Is the app installed?
+      #
+      # @return [Boolean] true or false
+      def app_installed?(bundle_id)
+        abstract_method!
+      end
+
+      # Install the app or ipa.
+      #
+      # If the app is already installed, it will be reinstalled from disk;
+      # no version check is performed.
+      #
+      # App data is never preserved.  If you want to preserve the app data,
+      # call `ensure_app_installed`.
+      #
+      # Possible return values:
+      #
+      # * :reinstalled => app was installed, but app data was not preserved.
+      # * :installed => app was not installed.
+      #
+      # @raise [InstallError] If app was not installed.
+      # @return [Symbol] A keyword describing the action that was performed.
+      def install_app(app_or_ipa)
+        abstract_method!
+      end
+
+      # Uninstall the app with bundle_id.
+      #
+      # App data is never preserved.  If you want to install a new version of
+      # an app and preserve app data (upgrade testing), call
+      # `ensure_app_installed`.
+      #
+      # Possible return values:
+      #
+      # * :nothing => app was not installed
+      # * :uninstall => app was uninstalled
+      #
+      # @raise [UninstallError] If the app cannot be uninstalled, usually
+      #   because it is a system app.
+      # @return [Symbol] A keyword that describes what action was performed.
+      def uninstall_app(bundle_id)
+        abstract_method!
+      end
+
+      # Ensures the app is installed and ensures that app is not stale by
+      # asking if the version of installed app is different than the version
+      # of the app or ipa on disk.
+      #
+      # The concrete implementation needs to check the CFBundleVersion and
+      # the CFBundleShortVersionString.  If either are different, then the
+      # app should be reinstalled.
+      #
+      # If possible, the app data should be preserved across reinstallation.
+      #
+      # Possible return values:
+      #
+      # * :nothing => app was already installed and versions matched.
+      # * :upgraded => app was stale; newer version from disk was installed and
+      #                app data was preserved.
+      # * :reinstalled => app was stale; newer version from disk was installed,
+      #                   but app data was not preserved.
+      # * :installed => app was not installed.
+      #
+      # @raise [InstallError] If the app could not be installed.
+      # @raise [UninstallError] If the app could not be uninstalled.
+      #
+      # @return [Symbol] A keyword that describes the action that was taken.
+      def ensure_app_installed(app_or_ipa)
+        abstract_method!
```

Comment:
After a conversation with Tobias, I think I should rename this method because it overloads the Calabash 2.0 concept of `ensure_app_installed`.

The rest of the API satisfies the requirements of Calabash 2.0.  This is a _new_ behavior that is not yet supported in Calabash 2.0: upgrade testing.